govcd/cse_util.go

```diff
@@ -0,0 +1,919 @@
+package govcd
+
+import (
+	_ "embed"
+	"encoding/json"
+	"fmt"
+	semver "github.com/hashicorp/go-version"
+	"github.com/vmware/go-vcloud-director/v2/types/v56"
+	"github.com/vmware/go-vcloud-director/v2/util"
+	"net/url"
+	"regexp"
+	"sort"
+	"strconv"
+	"strings"
+	"time"
+)
+
+// getCseComponentsVersions gets the versions of the subcomponents that are part of Container Service Extension.
+func getCseComponentsVersions(cseVersion semver.Version) (*cseComponentsVersions, error) {
+	v41, _ := semver.NewVersion("4.1.0")
+	v42, _ := semver.NewVersion("4.2.0")
+	v43, _ := semver.NewVersion("4.3.0")
+
+	if cseVersion.GreaterThanOrEqual(v41) && cseVersion.LessThan(v42) {
+		return &cseComponentsVersions{
+			VcdKeConfigRdeTypeVersion: "1.1.0",
+			CapvcdRdeTypeVersion:      "1.2.0",
+			CseInterfaceVersion:       "1.0.0",
+		}, nil
+	} else if cseVersion.GreaterThanOrEqual(v42) && cseVersion.LessThan(v43) {
+		return &cseComponentsVersions{
+			VcdKeConfigRdeTypeVersion: "1.1.0",
+			CapvcdRdeTypeVersion:      "1.3.0",
+			CseInterfaceVersion:       "1.0.0",
+		}, nil
+	}
+	return nil, fmt.Errorf("the Container Service Extension version '%s' is not supported", cseVersion.String())
+}
+
+// cseConvertToCseKubernetesClusterType takes a generic RDE that must represent an existing CSE Kubernetes cluster,
+// and transforms it to an equivalent CseKubernetesCluster object that represents the same cluster, but
+// it is easy to explore and consume. If the input RDE is not a CSE Kubernetes cluster, this method
+// will obviously return an error.
+//
```

Comment:
Given how large the function is and the warning below, would be nice to have a bigger description here on of how it does it.

Reply:
Done, thanks